### DISCUSSION TITLE
`consult-git-grep` and `pathspec`

### DISCUSSION THREAD

**Discussion by ErnestKz** [minad/consult#1036] [Ideas]
Summary and extracts of the discussion at: https://www.reddit.com/r/emacs/comments/1crl9tr/consultgitgrep_doesnt_honor_pathspec/

Currently, the way `consult--git-grep-make-builder` constructs the command, it appends paths to-be-searched to end end of it.

This interferes with the user-provided `pathspec` which filters what files should be matched.

For example: `#hello -- -- :*.txt` turns into `git grep "hello" -- :*.txt .`, notice the `.` at the end.

The work-around is to pass nil as the paths.

```elisp
(advice-add
'consult--git-grep-make-builder :around
(lambda (fn &rest args) (apply fn '(nil))))
```

The only downside AFIK is that you will not be able to specify multiple worktrees to do the git-grep on. But for those use-cases, I guess they can be somewhat covered with a `consult-grep`.

Chances are, if you want this `consult-git-grep` workaround, then you might want to be aware of `consult-grep`s `--include` behaviour:
- https://github.com/minad/consult/discussions/1018#discussion-6700710
- the takeaway being, add `(setq grep-find-ignored-files nil)` in order for `grep` to have the more expected `--include` behavoiur.

**Comment by minad** [comment]
Couldn't we quite easily detect if a pathspec is present and then construct the command line differently? Or would this break other things or maybe it is not robust?

**Reply by ErnestKz** [reply]
That's a nice suggestion. I think the command builder could detect the second `--`, ,and insert the `[<tree> ...]` to precede it:
```
git grep [-a | --text] [-I] [--textconv] [-i | --ignore-case] [-w | --word-regexp]
	   [-v | --invert-match] [-h|-H] [--full-name]
	   [-E | --extended-regexp] [-G | --basic-regexp]
	   [-P | --perl-regexp]
	   [-F | --fixed-strings] [-n | --line-number] [--column]
	   [-l | --files-with-matches] [-L | --files-without-match]
	   [(-O | --open-files-in-pager) [<pager>]]
	   [-z | --null]
	   [ -o | --only-matching ] [-c | --count] [--all-match] [-q | --quiet]
	   [--max-depth <depth>] [--[no-]recursive]
	   [--color[=<when>] | --no-color]
	   [--break] [--heading] [-p | --show-function]
	   [-A <post-context>] [-B <pre-context>] [-C <context>]
	   [-W | --function-context]
	   [(-m | --max-count) <num>]
	   [--threads <num>]
	   [-f <file>] [-e] <pattern>
	   [--and|--or|--not|(|)|-e <pattern>…​]
	   [--recurse-submodules] [--parent-basename <basename>]
	   [ [--[no-]exclude-standard] [--cached | --untracked | --no-index] | <tree>…​]
	   [--] [<pathspec>…​]
```
- https://git-scm.com/docs/git-grep

**Reply by ErnestKz** [reply]
Looks like, for full completeness, if there is a `[--[no-]exclude-standard]`, then `<tree>...` should not be there.
